src/MatrixClientPeg.ts

```diff
@@ -268,11 +269,45 @@ class _MatrixClientPeg implements IMatrixClientPeg {
             cryptoCallbacks: {},
         };
 
+        if (creds.olmAccount) {
+            console.log("got a dehydrated account");
+            const pickleKey = creds.pickleKey || "DEFAULT_KEY";
+            opts.deviceToImport = {
+                olmDevice: {
+                    pickledAccount: creds.olmAccount.pickle(pickleKey),
+                    sessions: [],
+                    pickleKey: pickleKey,
+                },
+                userId: creds.userId,
+                deviceId: creds.deviceId,
+            };
+            creds.olmAccount.free();
+        } else {
+            opts.userId = creds.userId;
+            opts.deviceId = creds.deviceId;
+        }
+
         // These are always installed regardless of the labs flag so that
         // cross-signing features can toggle on without reloading and also be
         // accessed immediately after login.
         Object.assign(opts.cryptoCallbacks, crossSigningCallbacks);
 
+        // set dehydration key after cross-signing gets set up -- we wait until
+        // cross-signing is set up because we want to cross-sign the dehydrated
+        // device
+        const origGetSecretStorageKey = opts.cryptoCallbacks.getSecretStorageKey
+        opts.cryptoCallbacks.getSecretStorageKey = async (keyinfo, ssssItemName) => {
```

Comment:
Rather than wrapping functions to achieve this, I think it would be easier to follow if an event was emitted for the thing you're waiting for (cross-signing setup), and then listen for that to trigger this work.